web/src/engine/predictive-text/worker-thread/src/main/correction/classical-calculation.ts

```diff
@@ -272,78 +324,210 @@ export class ClassicalDistanceCalculation<TUnit = string, TInput extends EditTok
+      }
+      const tier1 = b.maxMS - a.maxMS;
+      if(tier1 != 0) {
+        return tier1;
+      }
+      return b.maxMST - a.maxMST;
+    });
+
+    return properties.map((p) => p.result);
+  };
+
   /**
    * Determines the edit path used to obtain the optimal cost, distinguishing between zero-cost
    * substitutions ('match' operations) and actual substitutions.
    * @param row
    * @param col
    */
-  public editPath(row: number = this.inputSequence.length - 1, col: number = this.matchSequence.length - 1): EditOperation[] {
-    let currentCost = this.getCostAt(row, col);
-    let ops: EditOperation[] = null;
-    let parent: [number, number] = null;
-
-    let insertParentCost = this.getCostAt(row, col-1);
-    let deleteParentCost = this.getCostAt(row-1, col);
-    let substitutionParentCost = this.getCostAt(row-1, col-1);
-    let [lastInputIndex, lastMatchIndex] = ClassicalDistanceCalculation.getTransposeParent(this, row, col);
+  private _editPath(
+    row: number = this.inputSequence.length - 1,
+    col: number = this.matchSequence.length - 1
+  ): EditTuple<TUnit>[][] {
+    const currentCost = this.getCostAt(row, col);
+    const validPaths: EditTuple<TUnit>[][] = [];
+
+    const tryPath = (row: number, col: number, ops: EditTuple<TUnit>[]) => {
+      // Recursively build the edit path.
+      let results: EditTuple<TUnit>[][];
+      if(row >= 0 && col >= 0) {
+        results = this._editPath(row, col);
+      } else {
+        results = [[]];
+        const result = results[0];
+        for(let r = 0; r <= row; r++) {
+          // There are initial deletions.
+          result.push({
+            op: 'delete',
+            input: this.inputSequence[r].key
+          });
+        }
+        for(let c = 0; c <= col; c++) {
+          // There are initial insertions.
+          result.push({
+            op: 'insert',
+            match: this.matchSequence[c].key
+          });
+        }
+      }
+
+      // If null, there must not be any valid results
+      if(results) {
+        // ... also, if the array's empty.
+        results.forEach(r => validPaths.push(r.concat(ops)));
+      }
+    }
+
+    const [lastInputIndex, lastMatchIndex] = ClassicalDistanceCalculation.getTransposeParent(this, row, col);
     if(lastInputIndex >= 0 && lastMatchIndex >= 0) {
       // OK, a transposition source is quite possible.  Still need to do more vetting, to be sure.
       let expectedCost = 1;
 
       // This transposition includes either 'transpose-insert' or 'transpose-delete' operations.
-      ops = ['transpose-start']; // always needs a 'start'.
+      let i = row;
+      let m = col;
+      let ops: EditTuple<TUnit>[] = [];
+
       if(lastInputIndex != row-1) {
-        let count = row - lastInputIndex - 1;
-        ops = ops.concat( Array(count).fill('transpose-delete') );
-        expectedCost += count;
+        let count = row - lastInputIndex;
+        ops.push({
+          op: 'transpose-start',
```

Comment:
It was a little tricky to get right, but even transposition paths are now visualized well.